pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,351 @@
+	}
+	if err != nil {
+		var httpErr httpError
+		if errors.As(err, &httpErr) {
+			level.Warn(logger).Log("msg", httpErr.message, "err", err)
+			http.Error(w, httpErr.message, httpErr.statusCode)
+			return
+		}
+
+		level.Error(logger).Log("msg", "an unexpected error occurred", "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (c *MultitenantCompactor) createBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "starting block upload")
+
+	exists, err := userBkt.Exists(ctx, path.Join(blockID.String(), block.MetaFilename))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to check existence of %s in object storage", block.MetaFilename))
+	}
+	if exists {
+		level.Debug(logger).Log("msg", "complete block already exists in object storage")
+		return httpError{
+			message:    "block already exists in object storage",
+			statusCode: http.StatusConflict,
+		}
+	}
+
+	meta, err := decodeMeta(r.Body, "request body")
+	if err != nil {
+		return httpError{
+			message:    "malformed request body",
+			statusCode: http.StatusBadRequest,
+		}
+	}
+
+	if err := c.sanitizeMeta(logger, tenantID, blockID, &meta); err != nil {
+		return err
+	}
+
+	return c.uploadMeta(ctx, logger, meta, blockID, tmpMetaFilename, userBkt)
+}
+
+// UploadBlockFile handles requests for uploading block files.
+//
+// It takes the mandatory query parameter "path", specifying the file's destination path.
+func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	if blockID == "" {
+		http.Error(w, "missing block ID", http.StatusBadRequest)
+		return
+	}
+	_, err := ulid.Parse(blockID)
+	if err != nil {
+		http.Error(w, "invalid block ID", http.StatusBadRequest)
+		return
+	}
+	pth := r.URL.Query().Get("path")
+	if pth == "" {
+		http.Error(w, "missing or invalid file path", http.StatusBadRequest)
+		return
+	}
+
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	logger := util_log.WithContext(ctx, c.logger)
+	logger = log.With(logger, "block", blockID)
+
+	if path.Base(pth) == block.MetaFilename {
+		http.Error(w, fmt.Sprintf("%s is not allowed", block.MetaFilename), http.StatusBadRequest)
+		return
+	}
+
+	if !rePath.MatchString(pth) {
+		http.Error(w, fmt.Sprintf("invalid path: %q", pth), http.StatusBadRequest)
+		return
+	}
+
+	if r.ContentLength == 0 {
+		http.Error(w, "file cannot be empty", http.StatusBadRequest)
+		return
+	}
+
+	userBkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
+
+	metaPath := path.Join(blockID, tmpMetaFilename)
+	exists, err := userBkt.Exists(ctx, metaPath)
```

Comment:
We may want to read temporary meta.json instead -- it checks for its existence, we could reuse code from `completeBlockUpload`, and we will need it to do checks later (OK to include checks in the next PR).

Reply:
@pstibrany so you're saying I should change this into decoding the JSON from this file instead, even though we don't do any checks?

Reply:
Maybe might as well postpone such a change until we make checks versus (the temporary) meta.json?

Reply:
OK.